producer/writer/consumer_service_writer.go

```diff
@@ -0,0 +1,308 @@
+	if ct == topic.Unknown {
+		return nil, errUnknownConsumptionType
+	}
+	router := newAckRouter(int(numShards))
+	w := &consumerServiceWriterImpl{
+		cs:              cs,
+		numShards:       numShards,
+		ps:              ps,
+		shardWriters:    initShardWriters(router, ct, numShards, mPool, opts),
+		opts:            opts,
+		logger:          opts.InstrumentOptions().Logger(),
+		dataFilter:      acceptAllFilter,
+		router:          router,
+		consumerWriters: make(map[string]consumerWriter),
+		closed:          false,
+		m:               newConsumerServiceWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+	}
+	w.processFn = w.process
+	return w, nil
+}
+
+func initShardWriters(
+	router ackRouter,
+	ct topic.ConsumptionType,
+	numberOfShards uint32,
+	mPool messagePool,
+	opts Options,
+) []shardWriter {
+	sws := make([]shardWriter, numberOfShards)
+	for i := range sws {
+		switch ct {
+		case topic.Shared:
+			sws[i] = newSharedShardWriter(uint32(i), router, mPool, opts)
+		case topic.Replicated:
+			sws[i] = newReplicatedShardWriter(uint32(i), numberOfShards, router, mPool, opts)
+		}
+	}
+	return sws
+}
+
+func (w *consumerServiceWriterImpl) Write(d producer.RefCountedData) error {
+	shard := d.Shard()
+	if shard >= w.numShards {
+		w.m.invalidShard.Inc(1)
+		return fmt.Errorf("could not write data for shard %d which is larger than max shard id %d", shard, w.numShards)
+	}
+	if d.Accept(w.dataFilter) {
+		w.shardWriters[shard].Write(d)
+		w.m.filteredPassed.Inc(1)
+	}
+	// It is not an error if the data does not pass the filter.
+	w.m.filterNotPassed.Inc(1)
+	return nil
+}
+
+func (w *consumerServiceWriterImpl) Init() {
+	updatableFn := func() (watch.Updatable, error) {
+		return w.ps.Watch()
+	}
+	getFn := func(updatable watch.Updatable) (interface{}, error) {
+		update, err := updatable.(placement.Watch).Get()
+		if err != nil {
+			w.m.placementError.Inc(1)
+			w.logger.Errorf("invalid placement update from kv, %v", err)
+			return nil, err
+		}
+		w.m.placementUpdate.Inc(1)
+		return update, nil
+	}
+	vOptions := watch.NewOptions().
+		SetInitWatchTimeout(w.opts.PlacementWatchInitTimeout()).
+		SetInstrumentOptions(w.opts.InstrumentOptions()).
+		SetNewUpdatableFn(updatableFn).SetGetUpdateFn(getFn).SetProcessFn(w.processFn)
+	w.value = watch.NewValue(vOptions)
+	if err := w.startWatch(); err == nil {
+		return
+	}
+	// Since the consumer service writer could potentially be added during
+	// runtime while producer is already running, in case we failed to get
+	// a placement watch, we will retry forever in the background to establish
+	// the watch.
+	retrier := retry.NewRetrier(w.opts.PlacementWatchRetryOptions().SetForever(true))
+	continueFn := func(int) bool {
+		return !w.isClosed()
+	}
+	go retrier.AttemptWhile(continueFn, func() error {
+		return w.startWatch()
+	})
+}
+
+func (w *consumerServiceWriterImpl) startWatch() error {
+	err := w.value.Watch()
+	if err != nil && runtime.IsCreateWatchError(err) {
```

Comment:
So if err is not `CreateWatchError`, we are returning nil? That seems weird.

Reply:
yeah, there is not much we can do when a consumer service being added during runtime does not have a placement, but maybe I can change the behavior during process start up to error on InitValueError as well.